producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,317 @@
+		opts = NewOptions()
+	}
+	w := &consumerWriterImpl{
+		encdec:          proto.NewEncodeDecoder(defaultConn, opts.EncodeDecoderOptions()),
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		router:          router,
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		lastResetNanos:  0,
+		closed:          false,
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newConsumerWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	w.connectFn = w.connectOnce
+	if err := w.resetWithConnectFn(w.connectFn); err != nil {
+		w.signalReset()
+	}
+	return w
+}
+
+// Write should fail fast so that the write could be tried on other
+// consumer writers that are sharing the data.
+func (w *consumerWriterImpl) Write(m proto.Marshaler) error {
+	w.encodeLock.Lock()
+	err := w.encdec.Encode(m)
+	w.encodeLock.Unlock()
+	if err != nil {
+		w.m.encodeError.Inc(1)
+		w.signalReset()
+	}
+	return err
+}
+
+func (w *consumerWriterImpl) signalReset() {
+	w.RLock()
+	lastResetNanos := w.lastResetNanos
+	w.RUnlock()
+	// Avoid resetting too frequent.
+	if w.nowFn().UnixNano() < lastResetNanos+w.resetDelayNanos {
+		return
+	}
+	select {
+	case w.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (w *consumerWriterImpl) Init() {
+	w.closeWG.Add(2)
+	go w.readAcksForever()
+	go w.resetConnectionForever()
+}
+
+func (w *consumerWriterImpl) readAcksForever() {
+	defer w.closeWG.Done()
+
+	var acks msgpb.Ack
+	for !w.isClosed() {
+		w.decodeLock.Lock()
+		err := w.encdec.Decode(&acks)
+		w.decodeLock.Unlock()
+		if err != nil {
+			w.m.decodeError.Inc(1)
+			w.signalReset()
+			// Adding some delay here to avoid this being retried in a tight loop
+			// when underlying connection is misbehaving.
+			time.Sleep(w.opts.AckErrorRetryDelay())
+			continue
+		}
+		for _, m := range acks.Metadata {
+			if err := w.router.Ack(metadataFromProto(m)); err != nil {
+				w.m.ackError.Inc(1)
+				w.logger.Errorf("could not ack metadata, %v", err)
+			}
+		}
+		// NB(cw) The proto needs to be cleaned up because the gogo protobuf
+		// unmarshalling will append to the underlying slice.
+		acks.Metadata = acks.Metadata[:0]
+	}
+}
+
+func (w *consumerWriterImpl) resetConnectionForever() {
+	defer w.closeWG.Done()
+
+	for {
+		select {
+		case <-w.resetCh:
+			w.resetWithConnectFn(w.connectWithRetry)
+		case <-w.doneCh:
+			return
+		}
+	}
+}
+
+func (w *consumerWriterImpl) resetWithConnectFn(fn connectFn) error {
+	w.m.resetConn.Inc(1)
+	conn, err := fn(w.addr)
+	if err != nil {
+		w.m.resetError.Inc(1)
+		w.logger.Errorf("could not connect to %s, %v", w.addr, err)
+		return err
+	}
+	w.reset(conn)
+	return nil
+}
+
+func (w *consumerWriterImpl) connectWithRetry(addr string) (net.Conn, error) {
+	continueFn := func(int) bool {
```

Comment:
Nit: Idiomatically this makes sense but it might be worth defining continueFn and fn outside of connectWithRetry so you're not continuously defining these functions  when this is called. Up to you though.

Reply:
I can't move `fn` outside as I need to access the conn variable here, I could do continueFn, but then the signature requires an integer parameter that only makes sense with the retrier so I'd rather keep it closer to the retrier, also this is not the hot path so I'm not worried about repeatedly defining these functions

Reply:
fwiw I think the function itself will only be defined once and the captured variables, for example `w` in this case, will essentially be passed as parameters. This requires that the captured variables are allocated on the heap but shouldn't have any additional overhead compared to a normal function call.